examples/idle.c

```diff
@@ -0,0 +1,171 @@
+}
+
+static const struct org_kde_kwin_idle_timeout_listener idle_timer_listener = {
+	.idle = handle_idle,
+	.resumed = handle_resume,
+};
+
+int parse_args(int argc, char *argv[]) {
+	int c;
+	while ((c = getopt(argc, argv, "c:hs:t:")) != -1) {
+		switch(c)
+		{
+			case 'c':
+				close_timeout = strtoul(optarg, NULL, 10);
+				break;
+			case 's':
+				simulate_activity_timeout = strtoul(optarg, NULL, 10);
+				break;
+			case 't':
+				timeout = strtoul(optarg, NULL, 10);
+				break;
+			case 'h':
+			case '?':
+				printf("Usage: %s [OPTIONS]\n", argv[0]);
+				printf("  -t seconds\t\t\tidle timeout in seconds\n");
+				printf("  -s seconds optional\t\tsimulate user activity after x seconds\n");
+				printf("  -c seconds optional\t\tclose program after x seconds\n");
+				printf("  -h\t\t\t\tthis help menu\n");
+				return 1;
+			default:
+				return 1;
+		}
+	}
+	return 0;
+}
+
+void *simulate_activity(void *data) {
+	sleep(simulate_activity_timeout);
+	fprintf(stdout, "simulate user activity\n");
+	struct thread_args *arg = data;
+	org_kde_kwin_idle_timeout_simulate_user_activity(arg->timer);
+	wl_display_roundtrip(arg->display);
+	return NULL;
+}
+
+void *close_program(void *data) {
+	sleep(close_timeout);
+	struct thread_args *arg = data;
+	org_kde_kwin_idle_timeout_release(arg->timer);
+	wl_display_roundtrip(arg->display);
+	fprintf(stdout, "close program\n");
+	run = 0;
+	return NULL;
+}
+
+int main(int argc, char *argv[]) {
+
+	if(parse_args(argc, argv) != 0) {
+		return -1;
+	}
+	if(timeout == 0) {
+		printf("idle timeout 0 is invalid\n");
+		return -1;
+	}
+
+	struct wl_display *display = wl_display_connect(NULL);
+	if (display == NULL) {
+		fprintf(stderr, "failed to create display\n");
+		return -1;
+	}
+
+	struct wl_registry *registry = wl_display_get_registry(display);
+	wl_registry_add_listener(registry, &registry_listener, NULL);
+	wl_display_dispatch(display);
+	wl_display_roundtrip(display);
+
+	if (idle_manager == NULL) {
+		fprintf(stderr, "display doesn't support idle protocol\n");
+		return -1;
+	}
+	if (seat== NULL) {
+		fprintf(stderr, "seat error\n");
+		return -1;
+	}
+	struct org_kde_kwin_idle_timeout *timer =
+		org_kde_kwin_idle_get_idle_timeout(idle_manager, seat, timeout * 1000);
+
+	if (timer == NULL) {
+		fprintf(stderr, "Could not create idle_timeout\n");
+		return -1;
+	}
+
+	pthread_t t1, t2;
+	struct thread_args arg = {
+		.timer = timer,
+		.display = display,
+	};
+
+	if (simulate_activity_timeout != 0 && simulate_activity_timeout < close_timeout) {
+		if(pthread_create(&t1, NULL, &simulate_activity, (void *)&arg) != 0)
+			return -1;
+	}
+	if (close_timeout != 0) {
+		if (pthread_create(&t2, NULL, &close_program, (void *)&arg) != 0) {
+			if (simulate_activity_timeout != 0) {
+				pthread_cancel(t1);
+			}
+			return -1;
+		}
+	}
+
+	org_kde_kwin_idle_timeout_add_listener(timer, &idle_timer_listener, timer);
+	fprintf(stdout, "waiting\n");
+	while (run) {
+		wl_display_roundtrip(display);
+		sleep(1);
```

Comment:
This isn't the way to run the main loop. I think you want to use `wl_display_run()`.

Reply:
`wl_display_run` is part of the wayland server API. This example is suppose to be a wayland client. I'll rework the example to use `wl_display_dispatch`  and get rid of the `sleep`